### PR TITLE
aes: initial implementation of most aes instructions

### PR DESCRIPTION
**Pull request by Vineg**
Hi!

Regarding issue #45

I've adapted AES encryption/decryption instructions using algorithm from this project https://github.com/dhuertas/AES
The code is straight forward with a very basic optimization using a lookup table for the most time consuming function.
The only modification I did to the original implementation is throwing away 90% of unused lookup table data, leaving only 2 KB of the initial 60 KB.

### CODE REVIEW

File: simde/x86/aes.h [hunk 1]
```diff
@@ -0,0 +1,603 @@
...
+    a <<= 1;
+    if (hbs) a ^= 0x1b; // 0000 0001 0001 1011
+    b >>= 1;
+  }
+
+  return (uint8_t)p;
+}
+ */
+
+static uint8_t softaes_gmult_lookup_table[8][256] = {
```
Comment (mr-c) [inline simde/x86/aes.h:65]: Can these private variables and functions have a `_simde_` added to the beginning of their names to reduce the change of colliding with users' code?

File: simde/x86/aes.h [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+
+SIMDE_FUNCTION_ATTRIBUTES
+simde__m128i simde_mm_aesenclast_si128(simde__m128i a, simde__m128i round_key) {
+  #if defined(SIMDE_X86_AES_NATIVE)
+  return _mm_aesenclast_si128(a, round_key);
+  #else
+  simde__m128i result;
+  softaes_enc((uint8_t *) &a, (uint8_t *) &result, (uint8_t *) &round_key, 1);
+  return result;
+  #endif
```
Comment (mr-c) [inline simde/x86/aes.h:549]: Please indent like in the other functions, thanks

```suggestion
  #if defined(SIMDE_X86_AES_NATIVE)
    return _mm_aesenclast_si128(a, round_key);
  #else
    simde__m128i result;
    softaes_enc((uint8_t *) &a, (uint8_t *) &result, (uint8_t *) &round_key, 1);
    return result;
  #endif
```

File: simde/x86/aes.h [hunk 3]
```diff
@@ -0,0 +1,603 @@
...
+/*
+ * Transformation in the Cipher that takes all of the columns of the 
+ * State and mixes their data (independently of one another) to 
+ * produce new columns.
+ */
+SIMDE_FUNCTION_ATTRIBUTES
+void softaes_mix_columns(uint8_t *state) {
+
+  int Nb = softaes_Nb;
+  uint8_t k[] = {0x02, 0x01, 0x01, 0x03}; // a(x) = {02} + {01}x + {01}x2 + {03}x3
```
Comment (mr-c) [inline simde/x86/aes.h]: remove unused variable
```suggestion
```

File: simde/x86/aes.h [hunk 4]
```diff
@@ -0,0 +1,603 @@
...
+
+/*
+ * Transformation in the Inverse Cipher that is the inverse of 
+ * MixColumns().
+ */
+SIMDE_FUNCTION_ATTRIBUTES
+void softaes_inv_mix_columns(uint8_t *state) {
+
+  int Nb = softaes_Nb;
+  uint8_t k[] = {0x0e, 0x09, 0x0d, 0x0b}; // a(x) = {0e} + {09}x + {0d}x2 + {0b}x3
```
Comment (mr-c) [inline simde/x86/aes.h]:
```suggestion
```

File: simde/x86/aes.h [hunk 5]
```diff
@@ -0,0 +1,603 @@
...
+  #if defined(SIMDE_X86_AES_NATIVE)
+  return _mm_aesimc_si128(a);
+  #else
+  simde__m128i result;
+
+  uint8_t *in = (uint8_t *) &a;
+  uint8_t *out = (uint8_t *) &result;
+
+  int Nb = softaes_Nb;
+  uint8_t k[] = {0x0e, 0x09, 0x0d, 0x0b}; // a(x) = {0e} + {09}x + {0d}x2 + {0b}x3
```
Comment (mr-c) [inline simde/x86/aes.h]:
```suggestion
```

File: simde/x86/aes.h [hunk 6]
```diff
@@ -0,0 +1,603 @@
...
+
+  int Nb = softaes_Nb;
+  uint8_t i, k, s, tmp;
+
+  for (i = 1; i < 4; i++) {
+    s = 0;
+    while (s < i) {
+      tmp = state[Nb*i+Nb-1];
+
+      for (k = Nb-1; k > 0; k--) {
```
Comment (mr-c) [inline simde/x86/aes.h:403]: [`error: implicit conversion loses integer precision: 'int' to 'uint8_t' (aka 'unsigned char')`](https://github.com/simd-everywhere/simde/actions/runs/6497247919/job/17645921814?pr=1072#step:8:1095)

File: simde/x86/aes.h [hunk 7]
```diff
@@ -0,0 +1,603 @@
...
+simde__m128i simde_mm_aesenc_si128(simde__m128i a, simde__m128i round_key) {
+  #if defined(SIMDE_X86_AES_NATIVE)
+    return _mm_aesenc_si128(a, round_key);
+  #else
+    simde__m128i result;
+    softaes_enc((uint8_t *) &a, (uint8_t *) &result, (uint8_t *) &round_key, 0);
+    return result;
+  #endif
+}
+#if defined(SIMDE_X86_AES_ENABLE_NATIVE_ALIASES)
```
Comment (mr-c) [inline simde/x86/aes.h:523]: Please define `SIMDE_X86_AES_ENABLE_NATIVE_ALIASES` in `simde/simde-features.h` (around line 568)

**Review by mr-c — CHANGES_REQUESTED** [review]
Thank you very much for this!

Will you be adding optimized implementations for non-x86 systems as well?

**Comment by mr-c** [comment]
Please also rebase upon https://github.com/simd-everywhere/simde/tree/master to get the latest changes.

**Comment by mr-c** [comment]
I see that the neon implementations (with and without crypto extensions) at https://github.com/DLTcollab/sse2neon/pull/6 are also MIT licensed

See also
- https://github.com/DLTcollab/sse2neon/pull/63
- https://github.com/DLTcollab/sse2neon/pull/559
- https://github.com/DLTcollab/sse2neon/pull/75

File: simde/x86/aes.h [hunk 8]
```diff
@@ -0,0 +1,605 @@
...
+  }
+}
+
+SIMDE_FUNCTION_ATTRIBUTES
+simde__m128i simde_mm_aesenc_si128(simde__m128i a, simde__m128i round_key) {
+  #if defined(SIMDE_X86_AES_NATIVE)
+    return _mm_aesenc_si128(a, round_key);
+  #else
+    simde__m128i result;
+    softaes_enc((uint8_t *) &a, (uint8_t *) &result, (uint8_t *) &round_key, 0);
```
Comment (mr-c) [inline simde/x86/aes.h:519]: Please use `HEDLEY_REINTERPRET_CAST` or `HEDLEY_STATIC_CAST` as appropriate; this helps maintain C & C++ compatibility.

**Comment by mr-c** [comment]
Whoops, I should have said `simde_x_`, not `_simde_` as the prefix

File: simde/x86/aes.h [hunk 9]
```diff
@@ -0,0 +1,605 @@
...
+
+SIMDE_FUNCTION_ATTRIBUTES
+simde__m128i simde_mm_aesimc_si128(simde__m128i a) {
+  #if defined(SIMDE_X86_AES_NATIVE)
+    return _mm_aesimc_si128(a);
+  #else
+    simde__m128i result;
+
+    uint8_t *in = (uint8_t *) &a;
+    uint8_t *out = (uint8_t *) &result;
```
Comment (mr-c) [inline simde/x86/aes.h:577]: needs CAST macros

File: simde/x86/aes.h [hunk 10]
```diff
@@ -0,0 +1,605 @@
...
+      for (i = 0; i < 4; i++) {
+        out[Nb*j+i] = res[i];
+      }
+    }
+
+    return result;
+  #endif
+}
+#if defined(SIMDE_X86_AES_ENABLE_NATIVE_ALIASES)
+  #define _mm_aesimc_si128(a, b) simde_mm_aesimc_si128(a, b)
```
Comment (mr-c) [inline simde/x86/aes.h:600]:
```suggestion
  #define _mm_aesimc_si128(a) simde_mm_aesimc_si128(a)
```

**Comment by Vineg** [comment]
Michael, Thanks for review!

I am not going to add optimizations for other platforms. It is very interesting, but I don't currently have free time for it :)

**Review by mr-c — APPROVED** [review]

**Pull request merged**